full-service/src/db/mod.rs

```diff
@@ -68,4 +98,259 @@ impl WalletDb {
+        let cur_password_hash = self.password_hash.lock()?;
+        Ok(cur_password_hash.clone())
+    }
+
+    pub fn is_unlocked(&self) -> Result<bool, WalletDbError> {
+        Ok(!self.password_hash.lock()?.is_empty())
+    }
+
+    pub fn unlock(&self, password_hash: &[u8]) -> Result<(), WalletDbError> {
+        // No need to check db state if we're already unlocked
+        if self.is_unlocked()? {
+            return Ok(());
+        }
+
+        let conn = self.get_conn()?;
+        // Check whether encrypted, and if so, then attempt to unlock
+        match EncryptionIndicator::get_encryption_state(&conn)? {
+            EncryptionState::Empty => {
+                log::info!(
+                    self.logger,
+                    "DB has never been locked. Please call set_password to enable encryption."
+                );
+                return Err(WalletDbError::SetPassword);
+            }
+            EncryptionState::Encrypted => {
+                log::debug!(self.logger, "DB is locked. Verifying password.");
+                // Attempt to decrypt the test value to confirm if password is correct
+                let expected_val = Self::encrypt(ENCRYPTION_VERIFICATION_VAL, password_hash)?;
+                if EncryptionIndicator::verify_password(&expected_val, &conn)? {
+                    // Store password hash in memory
+                    let mut cur_password_hash = self.password_hash.lock()?;
+                    *cur_password_hash = password_hash.to_vec();
+                } else {
+                    return Err(WalletDbError::PasswordFailed);
+                }
+            }
+            EncryptionState::Unencrypted => {
+                log::info!(
+                    self.logger,
+                    "DB is unencrypted. Please call set_password to enable encryption."
+                );
+            }
+        }
+        Ok(())
+    }
+
+    pub fn change_password(
+        &self,
+        old_password_hash: &[u8],
+        new_password_hash: &[u8],
+    ) -> Result<(), WalletDbError> {
+        let conn = self.get_conn()?;
+        // Check whether encrypted, and if so, then verify old password. For any other state, set_password to new.
+        match EncryptionIndicator::get_encryption_state(&conn)? {
+            EncryptionState::Empty => {
+                log::info!(
+                    self.logger,
+                    "Database has never been locked. Setting password with new password."
+                );
+                self.set_password_hash(new_password_hash, &conn)?;
+            }
+            EncryptionState::Encrypted => {
+                log::debug!(self.logger, "Database is locked. Verifying old password.");
+                // Attempt to decrypt the test value to confirm if password is correct
+                let expected_val = Self::encrypt(ENCRYPTION_VERIFICATION_VAL, old_password_hash)?;
+                if EncryptionIndicator::verify_password(&expected_val, &conn)? {
```

Comment:
What happens to existing accounts when the password is changed?

Reply:
That's handled at the service layer - see `PasswordService::change_password`